File: python-algorithm/ds_4_linkedlist/linkedlist_1_vertical_calculation/L445-m.cpp
//
// Created by LiHeng on 2020/7/19.
//

#include <iostream>
#include <stack>
using namespace std;

//  Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

//题目要求不能修改原来的链表，所以就没法反转链表采用leetcode 2的做法
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* dummy = new ListNode(-1);
        stack<int> stk1 = buildStack(l1);
        stack<int> stk2 = buildStack(l2);
        int carry = 0;
        while(!stk1.empty() || !stk2.empty()){
            int val1 = 0;
            if(!stk1.empty()){
                val1 = stk1.top();
                stk1.pop();
            }
            int val2 = 0;
            if(!stk2.empty()){
                val2 = stk2.top();
                stk2.pop();
            }
            int tmp = val1 + val2 + carry;
            // 相当于不停的在dummy后面插入新的节点
            ListNode* node = new ListNode(tmp%10);
            node->next = dummy->next;
            dummy->next = node;
            carry = tmp/10;
        }
        //处理最后一次进位
        if(carry > 0){
            ListNode* node = new ListNode(1);
            node->next = dummy->next;
            dummy->next = node;
        }
        return dummy->next;
    }

    stack<int> buildStack(ListNode* l){
        stack<int> res;
        while(l != NULL){
            res.push(l->val);
            l = l->next;
        }
        return res;
    }
};

File: python-algorithm/ds_4_linkedlist/linkedlist_1_vertical_calculation/L445-m.py
#--coding:utf-8--


# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        dummy = ListNode(-1)
        cur = dummy
        tmp_l1, tmp_l2 = [], []
        while l1 != None:
            tmp_l1.append(l1.val)
            l1 = l1.next
        while l2 != None:
            tmp_l2.append(l2.val)
            l2 = l2.next
        carry = 0
        while tmp_l1 != [] or tmp_l2 != []:
            tmp_sum = 0
            if tmp_l1 != []:
                tmp_sum += tmp_l1[-1]
                tmp_l1.pop()
            if tmp_l2 != []:
                tmp_sum += tmp_l2[-1]
                tmp_l2.pop()
            tmp_sum += carry
            # 链表头插法的应用
            tmp_node = ListNode(tmp_sum % 10)
            tmp_node.next = cur.next
            cur.next = tmp_node
            carry = tmp_sum // 10  # python整除(取模)用// c++是/
        if carry:
            tmp_node = ListNode(1)
            tmp_node.next = cur.next
            cur.next = tmp_node
        return dummy.next


if __name__ == "__main__":
    ln1 = ListNode(7)
    p = ln1
    for i in [2,4,3]:
        p.next = ListNode(i)
        p = p.next

    ln2 = ListNode(5)
    q = ln2
    for i in [6,4]:
        q.next = ListNode(i)
        q = q.next

    res = Solution().addTwoNumbers(ln1, ln2)
    while res != None:
        print(res.val)
        res = res.next
